hoomd/tuner/balance.py

```diff
@@ -0,0 +1,97 @@
+    roughly :math:`I-1.0`, so that if the largest imbalance is 1.4, then the
+    user can expect a roughly 40% speedup in the simulation. This is of course
+    an estimate that assumes that all algorithms are roughly linear in
+    :math:`N`, all GPUs are fully occupied, and the simulation is limited by the
+    speed of the slowest processor. It also assumes that all particles roughly
+    equal. If you have a simulation where, for example, some particles have
+    significantly more pair force neighbors than others, this estimate of the
+    load imbalance may not produce the optimal results.
+
+    A load balancing adjustment is only performed when the maximum load
+    imbalance exceeds a *tolerance*. The ideal load balance is 1.0, so setting
+    *tolerance* less than 1.0 will force an adjustment every *period*. The load
+    balancer can attempt multiple iterations of balancing every *period*, and up
+    to *maxiter* attempts can be made. The optimal values of *period* and
+    *maxiter* will depend on your simulation.
+
+    Load balancing can be performed independently and sequentially for each
+    dimension of the simulation box. A small performance increase may be
+    obtained by disabling load balancing along dimensions that are known to be
+    homogeneous.  For example, if there is a planar vapor-liquid interface
+    normal to the :math:`z` axis, then it may be advantageous to disable
+    balancing along :math:`x` and :math:`y`.
+
+    In systems that are well-behaved, there is minimal overhead of balancing
+    with a small *period*. However, if the system is not capable of being
+    balanced (for example, due to the density distribution or minimum domain
+    size), having a small *period* and high *maxiter* may lead to a large
+    performance loss. In such systems, it is currently best to either balance
+    infrequently or to balance once in a short test run and then set the
+    decomposition statically in a separate initialization.
+
+    Balancing is ignored if there is no domain decomposition available (MPI is
+    not built or is running on a single rank).
+    """
+
+    def __init__(self, trigger, x=False, y=False, z=False, tolerance=1.02,
+                 max_iterations=1):
+        defaults = dict(x=x, y=y, z=z, tolerance=tolerance,
+                        max_iterations=max_iterations, trigger=trigger)
+        self._param_dict = ParameterDict(
+            x=bool, y=bool, z=bool, max_iterations=int, tolerance=float,
+            trigger=OnlyType(Trigger, preprocess=trigger_preprocessing))
+        self._param_dict.update(defaults)
+
+    def attach(self, simulation):
+        if simulation.device.mode == 'gpu':
+            cpp_cls = getattr(_hoomd, 'LoadBalancerGPU')
+        else:
+            cpp_cls = getattr(_hoomd, 'LoadBalancer')
+        self._cpp_obj = cpp_cls(simulation.state._cpp_sys_def,
+                                simulation.device.decomposition._cpp_obj,
```

Comment:
We can complete this PR now before completing the domain decomposition API. Add a `getDomainDecomposition` method to `Communicator.h`. Then replace this with `simulation._cpp_sys.getCommunicator().getDomainDecomposition()`

The decomposition is passed here as a matter of plumbing to link up the decomposition with the balancer, as the decomposition is not part of the `SystemDefinition`.